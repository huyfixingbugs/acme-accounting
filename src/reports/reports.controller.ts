import { Controller, Get, Post, HttpCode } from '@nestjs/common';
import { ReportsService } from './reports.service';

@Controller('api/v1/reports')
export class ReportsController {
  constructor(private reportsService: ReportsService) {}

  @Get()
  report() {
    return {
      'accounts.csv': this.reportsService.state('accounts'),
      'yearly.csv': this.reportsService.state('yearly'),
      'fs.csv': this.reportsService.state('fs'),
    };
  }

  @Post()
  @HttpCode(201)
  generate() {
    const start = performance.now();

    this.reportsService.accounts();
    this.reportsService.yearly();
    this.reportsService.fs();

    return { message: `finished in ${((performance.now() - start) / 1000).toFixed(2)}` };
  }
}
