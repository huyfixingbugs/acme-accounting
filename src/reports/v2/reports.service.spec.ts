import { Test, TestingModule } from "@nestjs/testing";
import { ReportsServiceV2 } from "./reports.service";
import fs from "fs";
import path from "path";

describe('ReportsServiceV2', () => {
  let service: ReportsServiceV2;
  let outputDir = 'out';

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ReportsServiceV2],
    }).compile();

    service = module.get<ReportsServiceV2>(ReportsServiceV2);

    if (fs.existsSync(outputDir)) {
      fs.readdirSync(outputDir).forEach(file => {
        if (file.endsWith('.csv')) {
          fs.unlinkSync(path.join(outputDir, file));
        }
      });
    }
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should generate reports', async () => {
    await service.generate();
    const accountsOutput = fs.readFileSync(path.resolve(outputDir, 'accounts-v2.csv'), 'utf8');
    const yearlyOutput = fs.readFileSync(path.resolve(outputDir, 'yearly-v2.csv'), 'utf8');
    const fsOutput = fs.readFileSync(path.resolve(outputDir, 'fs-v2.csv'), 'utf8');
    
    expect(accountsOutput).toMatchSnapshot('accounts-v2.csv');
    expect(yearlyOutput).toMatchSnapshot('yearly-v2.csv');
    expect(fsOutput).toMatchSnapshot('fs-v2.csv');
  });

  it('should update status correctly', () => {
    const promise = service.generate();
    expect(service.state('accounts')).toBe('starting');
    expect(service.state('yearly')).toBe('starting');
    expect(service.state('fs')).toBe('starting');

    return promise.then(() => {
      expect(service.state('accounts')).toMatch(/finished in/);
      expect(service.state('yearly')).toMatch(/finished in/);
      expect(service.state('fs')).toMatch(/finished in/);
    });
  });
})
