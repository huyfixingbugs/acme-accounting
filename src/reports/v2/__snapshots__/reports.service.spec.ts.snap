// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsServiceV2 should generate reports: accounts-v2.csv 1`] = `
"Account,Balance
Fixed Assets,26113198.37
Tax Expense,39436232.24
Interest Expense,-1672198.58
Accounts Payable,21371673.27
Cash,-5952833.60
Accounts Receivable,-29817119.16
Unearned Revenue,5330650.00
Accrued Liabilities,-1082450.57
Salaries Expense,41094245.55
Common Stock,16645462.73
Loan Payable,2581401.93
Prepaid Expenses,-35602278.62
Cost of Goods Sold,6047460.55
Dividends Payable,-29637869.40
Retained Earnings,-17815584.94
Inventory,-19732455.29
Rent Expense,2710797.64
Sales Tax Payable,-37837055.25
Sales Revenue,14229202.79
Utilities Expense,16428280.46"
`;

exports[`ReportsServiceV2 should generate reports: fs-v2.csv 1`] = `
"Basic Financial Statement

Income Statement
Sales Revenue,14229202.79
Cost of Goods Sold,6047460.55
Salaries Expense,41094245.55
Rent Expense,2710797.64
Utilities Expense,16428280.46
Interest Expense,-1672198.58
Tax Expense,39436232.24
Net Income,-89815615.07

Balance Sheet
Assets
Cash,-5952833.60
Accounts Receivable,-29817119.16
Inventory,-19732455.29
Fixed Assets,26113198.37
Prepaid Expenses,-35602278.62
Total Assets,-64991488.30

Liabilities
Accounts Payable,21371673.27
Loan Payable,2581401.93
Sales Tax Payable,-37837055.25
Accrued Liabilities,-1082450.57
Unearned Revenue,5330650.00
Dividends Payable,-29637869.40
Total Liabilities,-39273650.02

Equity
Common Stock,16645462.73
Retained Earnings,-17815584.94
Retained Earnings (Net Income),-89815615.07
Total Equity,-90985737.28

Assets = Liabilities + Equity, -64991488.30 = -130259387.30"
`;

exports[`ReportsServiceV2 should generate reports: yearly-v2.csv 1`] = `
"Financial Year,Cash Balance
2019,195299.48
2020,-4944891.10
2021,12017491.75
2022,-18878324.27
2023,-15796925.76
2024,21454516.30"
`;
